### PR TITLE
Add fe plugin framework

### PR DESCRIPTION
**Pull request by Seaven**
For #2346
Add fe plugin framework

todo：
1. update plugin meta
2. add plugin install/uninstall statement

NOTICE: this CL update FeMetaVersion to 78

### CODE REVIEW

**Comment by imay** [comment]
@Seaven I think you should add some document or comment to help people review this.

**Comment by Seaven** [comment]
Support Fe Plugin framework:

do something:

* Add install plugin statement class
* Add uninstall plugin statement class

* Support Plugin manger
	* Support FE Plugin register method:  the fe built-in Plugin can register through the way.
	
	* Support FE Plugin dynamic install process
	  1. check Plugin has already install?
	  2. download Plugin file from remote source or copy from local source
	  3. extract Plugin's .zip 
	  4. read Plugin's plugin.properties, and check Plugin's Value
	  5. dynamic load .jar and init Plugin's main Class
	  6. invoke Plugin's init method
	  7. register Plugin into PluginMgr.
	
	* Support FE Plugin dynamic uninstall process
		* check Plugin has install?
		* invoke Plugin's close method
		* delete Plugin from PluginMgr

To do:

* install/uninstall Plugin should update meta
* for BE Plugin, need check valid through BE node. we need add rpc interface
* SQL support: install/unintall Plugin syntax
* SQL support: register the PluginMgr's install/uninstall process in QE
* load & init Plugin when the fe start.

File: fe/src/main/java/org/apache/doris/plugin/PluginInfo.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        }
+
+        final String name = propsMap.remove("name");
+        if (null == name || name.isEmpty()) {
+            throw new IllegalArgumentException(
+                    "property [name] is missing in [" + descriptor + "]");
+        }
+
+        final String description = propsMap.remove("description");
+        if (null == description) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/src/main/java/org/apache/doris/plugin/PluginInfo.java]: Yeah, is unnecessary check, will remove

File: fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        Path realPath = movePlugin();
+
+        plugin = dynamicLoadPlugin(realPath);
+
+        pluginInstallValid();
+
+        plugin.init(pluginInfo, pluginContext);
+    }
+
+    private boolean hasInstall() {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java]:
```suggestion
    private boolean hasInstalled() {
```

File: fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        // check already install
+        if (pluginInfo != null && installPath != null && Files.exists(installPath)) {
+            return true;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java]: Why not checking `plugin`？

File: fe/src/main/java/org/apache/doris/plugin/PluginMgr.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+     */
+    public PluginInfo installPlugin(InstallPluginStmt stmt) throws IOException, UserException {
+        PluginLoader pluginLoader = new DynamicPluginLoader(pluginDir, stmt.getPluginPath());
+
+        try {
+            PluginInfo info = pluginLoader.getPluginInfo();
+            // install plugin
+            pluginLoader.install();
+
+            PluginLoader checkLoader = plugins[info.getTypeId()].putIfAbsent(info.getName(), pluginLoader);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/plugin/PluginMgr.java]: Check the name first before install the plugin

**Review by morningman — CHANGES_REQUESTED** [review]
I think you can just make our current audit logger a builtin plugin

File: fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java (hunk 3 above)
Comment (Seaven) [reply fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java]: Pluginloader is used to initialize the plugin, so the `plugin` must be null before the installation is complete

**Comment by wutiangan** [comment]
@Seaven What problem does plugin used to solve? please give some example.
And how to use the plugin.

**Comment by Seaven** [comment]
> @Seaven What problem does plugin used to solve? please give some example.
> And how to use the plugin.

The plugin framework allows user to install or uninstall some custom features when the Doris is running, such as support different storages, data import formats or audit log（like audit_plugin_demo above）. And the framework is incomplete now, we will support more examples and documents later

File: fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+        getPluginInfo();
+
+        Path realPath = movePlugin();
+
+        plugin = dynamicLoadPlugin(realPath);
+
+        pluginInstallValid();
+
+        plugin.init(pluginInfo, pluginContext);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java]: the pluginContext is not initialized, and is `null`.

File: fe/src/main/java/org/apache/doris/plugin/PluginMgr.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+            if (plugins[info.getTypeId()].containsKey(info.getName())) {
+                throw new UserException("plugin " + info.getName() + " has install.");
+            }
+
+            // install plugin
+            pluginLoader.install();
+
+            if (plugins[info.getTypeId()].putIfAbsent(info.getName(), pluginLoader) != null) {
+                throw new UserException("plugin " + info.getName() + " has install.");
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/plugin/PluginMgr.java]:
```suggestion
                throw new UserException("plugin " + info.getName() + " has already been installed.");
```

File: fe/src/main/java/org/apache/doris/plugin/PluginMgr.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    /**
+     *
+     * Just load dynamic plugin
+     */
+    public void loadDynamicPlugin(PluginInfo info) throws IOException, UserException {
+        DynamicPluginLoader pluginLoader = new DynamicPluginLoader(Config.plugin_dir, info);
+
+        try {
+            // install plugin
+            pluginLoader.reload();
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/plugin/PluginMgr.java]: If we reload failed, an exception will be thrown, and the plugin info will not be added to the `plugins`. So in replay process, it means we lost this plugin.

So I think even the plugin is reloaded failed, we should always put it in the `plugins`.

And each time we get a plugin from plugin manager, we will check if it is loaded successfully.

File: fe/src/main/java/org/apache/doris/analysis/ShowPluginsStmt.java [hunk 8]
```diff
@@ -27,10 +27,13 @@
...
-                    .addColumn(new Column("Type", ScalarType.createVarchar(80)))
-                    .addColumn(new Column("Library", ScalarType.createVarchar(64)))
-                    .addColumn(new Column("License", ScalarType.createVarchar(80)))
+                    .addColumn(new Column("Type", ScalarType.createVarchar(10)))
+                    .addColumn(new Column("Description", ScalarType.createVarchar(200)))
+                    .addColumn(new Column("Version", ScalarType.createVarchar(20)))
+                    .addColumn(new Column("JavaVersion", ScalarType.createVarchar(20)))
+                    .addColumn(new Column("ClassName", ScalarType.createVarchar(64)))
+                    .addColumn(new Column("SoName", ScalarType.createVarchar(64)))
+                    .addColumn(new Column("Sources", ScalarType.createVarchar(200)))
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/ShowPluginsStmt.java]: How about adding a `Status` column to show that if the plugin is loaded successfully.

**Review by morningman — COMMENTED** [review]
Better to add a `README.md` under `fe_plugins` dir to explain:
1. What files should be with a plugin.
2. What the directory structure should look like.
3. How to compile the plugin.
4. How to deploy and install the plugin.

File: fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        return urls;
+    }
+
+    /**
+     * move plugin's temp install directory to Doris's PLUGIN_DIR
+     */
+    Path movePlugin() throws UserException, IOException {
+
+        if (tempPath == null || !Files.exists(tempPath)) {
+            throw new UserException("Install plugin " + pluginInfo.getName() + " failed, cause install path isn't "
```
Comment (wutiangan) [inline fe/src/main/java/org/apache/doris/plugin/DynamicPluginLoader.java]: cause --> becasue?

**Comment by morningman** [comment]
1. `Config.plugin_dir` should be created if it doesn' t exist, or exception throw:
    `java.nio.file.NoSuchFileException: /home/palo/output/fe/plugins/.install_7268007407478981557`

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**